crates/handlers/src/views/login.rs

```diff
@@ -207,76 +212,116 @@ pub(crate) async fn post(
-        .ok_or(FormError::InvalidCredentials)?;
-
-    // Check the rate limit
-    limiter.check_password(requester, &user).map_err(|e| {
-        tracing::warn!(error = &e as &dyn std::error::Error);
-        FormError::RateLimitExceeded
-    })?;
-
-    // And its password
-    let user_password = repo
-        .user_password()
-        .active(&user)
-        .await
-        .map_err(|_e| FormError::Internal)?
-        .ok_or(FormError::InvalidCredentials)?;
-
-    let password = Zeroizing::new(password.as_bytes().to_vec());
-
-    // Verify the password, and upgrade it on-the-fly if needed
-    let new_password_hash = password_manager
-        .verify_and_upgrade(
+    let mxid: String = homeserver.mxid(username);
+    let user = if let Ok(Some(rest_auth_provider)) = password_manager.get_rest_auth_provider() {
+        // If rest_auth_provider is enabled, authenticate via REST API
+        let authenticated =
+            authenticate_via_rest_api(mxid.clone(), password.to_owned(), rest_auth_provider)
+                .await
+                .map_err(|_err| FormError::InvalidCredentials)?;
+        if authenticated {
+            let user = repo
+                .user()
+                .find_by_username(username)
+                .await
+                .map_err(|_err| FormError::Internal)?
+                .filter(mas_data_model::User::is_valid);
+
+            let user = if let Some(user) = user {
+                // User found: proceed
+                user
+            } else {
+                // User not found while existing in the provider: create it
+                let homeserver_available = homeserver
+                    .is_localpart_available(&mxid)
+                    .await
+                    .map_err(|_err| FormError::Internal)?;
+                if !homeserver_available {
+                    return Err(FormError::Internal);
+                }
+
+                let new_user = repo
+                    .user()
+                    .add(&mut rng, clock, mxid.clone())
+                    .await
+                    .map_err(|_err| FormError::Internal)?;
+
+                // Replicate in Synapse
+                homeserver
+                    .provision_user(&ProvisionRequest::new(mxid.clone(), username.to_owned()))
+                    .await
+                    .unwrap();
+
+                new_user
+            };
+            // Update the password if needed
+            let result = password_manager
+                .hash(&mut rng, password.to_owned().into_bytes().into())
+                .await;
+            let (version, hashed_password) = match result {
+                Ok((version, hashed_password)) => (version, hashed_password),
+                Err(_err) => return Err(FormError::Internal),
+            };
+            repo.user_password()
+                .upsert(&mut rng, clock, &user, version, hashed_password)
+                .await
+                .map_err(|_err| FormError::Internal)?;
```

Comment:
I'm a little confused on the usefulness of getting that persisted in the database still? Why would we want that, if we're talking to an external service to get verify the passwords?